pkg/services/ngalert/api/forked_am.go

```diff
@@ -0,0 +1,183 @@
+package api
+
+import (
+	"fmt"
+
+	apimodels "github.com/grafana/alerting-api/pkg/api"
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+type ForkedAMSvc struct {
+	AmSvc, GrafanaSvc AlertmanagerApiService
+	DatasourceCache   datasources.CacheService
+}
+
+func NewForkedAM(datasourceCache datasources.CacheService, proxy, grafana AlertmanagerApiService) *ForkedAMSvc {
+	return &ForkedAMSvc{
+		AmSvc:           proxy,
+		GrafanaSvc:      grafana,
+		DatasourceCache: datasourceCache,
+	}
+}
+
+func (am *ForkedAMSvc) RouteCreateSilence(ctx *models.ReqContext, body apimodels.SilenceBody) response.Response {
+	t, err := backendType(ctx, am.DatasourceCache)
+	if err != nil {
+		return response.Error(400, err.Error(), nil)
+	}
+
+	switch t {
+	case apimodels.GrafanaBackend:
+		return am.GrafanaSvc.RouteCreateSilence(ctx, body)
+	case apimodels.AlertmanagerBackend:
+		return am.AmSvc.RouteCreateSilence(ctx, body)
+	default:
+		return response.Error(400, fmt.Sprintf("unexpected backend type (%v)", t), nil)
+	}
```

Comment:
Could we reduce this amount of boilerplate by using an interface instead? Something like:

```
	svc, err := backendType(ctx, am.DatasourceCache)
	if err != nil {
		return response.Error(400, err.Error(), nil)
	}

	return svc.RouteCreateSilence(ctx, body)
```

Would save a lot of code.